dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,243 @@
+    # created automatically when the authorization flow completes for the first
+    # time.
+    credentials_path = os.path.join(
+        os.path.expanduser('~'), self._CREDENTIALS_FILENAME)
+    lock = filelock.FileLock(credentials_path + '.lock')  # pylint: disable=abstract-class-instantiated
+    with lock:
+      if os.path.exists(credentials_path):
+        try:
+          credentials = Credentials.from_authorized_user_file(
+              credentials_path, self.SCOPES)
+        except ValueError as exception:
+          self.logger.warning(
+              'Unable to load credentials: {0:s}'.format(exception))
+          credentials = None
+
+      # If there are no (valid) credentials available, let the user log in.
+      if not credentials or not credentials.valid:
+        if credentials and credentials.expired and credentials.refresh_token:
+          credentials.refresh(Request())
+        else:
+          secrets_path = os.path.join(
+              os.path.expanduser('~'), self._CLIENT_SECRET_FILENAME)
+          if not os.path.exists(secrets_path):
+            error_message = (
+                'No OAuth application credentials available to access google '
+                'sheets. Please generate OAuth application credentials (see '
+                'https://developers.google.com/sheets/api/guides/authorizing) '
+                'and save them to {0:s}.').format(secrets_path)
+            self.ModuleError(error_message, True)
+          flow = InstalledAppFlow.from_client_secrets_file(
+              secrets_path, self.SCOPES)
+          credentials = flow.run_console()
+
+        # Save the credentials for the next run
+        with open(credentials_path, 'w') as token_file:
+          token_file.write(credentials.to_json())
+
+    return credentials
+
+  def _BuildSheetsResource(self, credentials: Credentials) -> discovery.Resource:
+    """Builds a Google Sheets resource object to use to request logs.
+
+    Args:
+      credentials: Google API credentials
+
+    Returns:
+      A resouce object for interacting with the Google Sheets API.
+    """
+    resource = discovery.build('sheets', 'v4', credentials=credentials)
+
+    return resource
+
+  def _ValidateSpreadSheetId(self, spreadsheet: str) -> str:
+    """Extract the spreadsheet id if the input is a URL and validate that the ID
+    is in the correct format.
+
+    Args:
+      spreadsheet: ID or URL of the sheetspread,
+
+    Returns:
+      spreadsheet ID
+    """
+    spreadsheet_match = re.search(r'.*?([01][0-9A-Za-z_-]{20,}).*',
+                                     spreadsheet)
+    if not spreadsheet_match:
+      self.ModuleError(
+          'spreadsheet id is not in the correct format {0:s}.'.format(
+              spreadsheet),
+          critical=True)
+
+    return spreadsheet_match.group(1)
+
+  def _ParseSheet(self, spreadsheet_id: str, sheet_title: str) -> DataFrame:
+    """Parse sheet content
```

Comment:
Can you be a bit more descriptive in the docstring here? What exactly are we parsing? What's the output of the parsing? e.g. "Parse sheet content and convert to Pandas dataframe"

Reply:
I am not realy parsing anything, I am extracting data. I changed the name of the method and doc string to reflect that